lib/tapioca/gem/listeners/foreign_constants.rb

```diff
@@ -0,0 +1,46 @@
+# typed: strict
+# frozen_string_literal: true
+
+module Tapioca
+  module Gem
+    module Listeners
+      class ForeignConstants < Base
+        extend T::Sig
+
+        include Runtime::Reflection
+
+        private
+
+        sig { override.params(event: ScopeNodeAdded).void }
+        def on_scope(event)
+          mixin = event.constant
+          return if Class === mixin # Classes can't be mixed into other constants
+
+          # There are cases where we want to process constants not declared by the current
+          # gem, i.e. "foreign constant". These are constants defined in another gem to which
+          # this gem is applying a mix-in. This pattern is especially common for gems that add
+          # behavior to Ruby standard library classes by mixing in modules to them.
+          #
+          # The way we identify these "foreign constants" is by asking the mixin tracker which
+          # constants have mixed in the current module that we are handling. We add all the
+          # constants that we discover to the pipeline to be processed.
+          Runtime::Trackers::Mixin.constants_with_mixin(mixin).each do |constant, locations|
+            next unless mixed_in_by_gem?(locations)
```

Comment:
@paracycle Here's the use of that method. As far as I could tell, we were never actually using the `mixin_type` from the tracker, so I've just been storing the constant and locations. Let me know if this looks alright!

Reply:
Yes, for this part of the work, this much is enough. 

However, later on we will also have to revert https://github.com/Shopify/tapioca/pull/668 at which point, we will have start deciding, for a given mixin and a given mixin type (`include`, `extend`, `prepend`), is that happening from inside this gem or not. At that point, we will need that logic.